format/ComputeIR.fbs

```diff
@@ -0,0 +1,348 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+include "Schema.fbs";
+
+namespace org.apache.arrow.flatbuf.computeir;
+
+/// Avoid use of org.apache.arrow.Buffer because it requires a
+/// sidecar block of bytes.
+table InlineBuffer {
+  // ulong is used to guarantee alignment and padding of `bytes` so that flatbuffers
+  // and other alignment sensitive blobs can be stored here
+  bytes: [ulong] (required);
+}
+
+/// An expression is one of
+/// - a Literal datum
+/// - a reference to a Field from a Relation
+/// - a call to a named function
+/// On evaluation, an Expression will have either array or scalar shape.
+union ExpressionImpl {
+  Literal, FieldRef, Call
+}
+
+table Expression {
+  // Ideally we'd simply have `union Expression { Literal, FieldRef, Call }`
+  // but not all generators support vectors of unions so we provide minimal
+  // indirection to support them.
+  impl: ExpressionImpl (required);
+}
+
+union Shape {
+  Array, Scalar
+}
+
+table Scalar {}
+
+table Array {
+  /// Number of slots.
+  length: long;
+}
+
+table Literal {
+  /// Shape of this literal.
+  ///
+  /// Note that this is orthogonal to type and refers to the number
+  /// of rows spanned by this Literal - a Literal may be Scalar shaped
+  /// with multiple "columns" if the type happens to be Struct.
+  shape: Shape (required);
+
+  /// The type of this literal. Field is used instead of Type to pick
+  /// up child fields, dictionary encoding, etc.
+  field: Field (required);
```

Comment:
It seems simpler to put the field/output type in `Expression` instead of replicating it across each type in `ExpressionImpl`

Reply:
SGTM